### ISSUE TITLE
cse.cluster.id is read/write

### ISSUE THREAD

**Issue by m1gu3lg** [vmware/container-service-extension#542]
I have noticed that all the metadata introduced by CSE 2.6beta1 in the vCD App are Read/Write. I understand this makes sense for some of them, since the tenant might choose to upgrade the cluster manually (i.e., not using CSE) and could want to update those elements manually (e.g. the K8s version). However, some attributes like cse.cluster.id are permanent and should be readonly. This would be really useful for us, since then we could consider them for accounting purposes. Could the cse.cluster.id metadata be readonly in CSE 2.6 GA?

**Comment by ejgarcia** [comment]
I agree: all vApp metadata is defined in the general domain, and it shouldn't be so, since this is basic for the proper working of the cluster that should not be accessible to the user. In particular, in "server_cli", the following sentence appears a couple of times:

`vapp.set_multiple_metadata(new_metadata_to_add)`

The default values from pyvcloud.vcd.vapp are:

`def set_multiple_metadata(self,
                              key_value_dict,
                              domain=MetadataDomain.GENERAL,
                              visibility=MetadataVisibility.READ_WRITE,
                              metadata_value_type=MetadataValueType.STRING):`

In contrast, for orgVDC, metadata is properly set:

`return ovdc.set_multiple_metadata(metadata, MetadataDomain.SYSTEM,
                                          MetadataVisibility.PRIVATE)`

Is there any reason why vApp metadata is made editable for the end-user?

**Comment by rocknes** [comment]
Hi,

There is a bug around the whole metadata workflow in vCD. I encountered it while adding support for metadata READ/WRITE in pyvcloud https://github.com/vmware/pyvcloud/blob/c4017e701bf6e170c8ac8a20de9e5b1fa666a6f1/pyvcloud/vcd/client.py#L513

GENERAL - READONLY is not a valid combination as per vCD logic :( Hence it's in READ/WRITE mode. And the reason the cluster metadata is in GENERAL domain is purely historical. With the next release of vCD after 10.1, we will move away from storing metadata on vApps. We will keep in consideration to fix the metadata visibility and editibility at that time.

Regards
Aritra Sen

**Comment by ejgarcia** [comment]
Hi, Aritra,

by looking at the code, I expected that changing to the proper SYSTEM domain was as simple as replacing the line to:

`vapp.set_multiple_metadata(new_metadata_to_add, MetadataDomain.SYSTEM, MetadataVisibility.READONLY`

I understand that it is not a matter of simplicity, but just that the metadata model will be changed in future versions. If you will not be storing metadata in vApps, what is the solution you are planning? I actually did find most useful to have that metadata there, in order to identify clusters for statistics and billing purposes.

Kind regards,

Emilio

**Comment by rocknes** [comment]
Hi Emilio,

vCD is coming up with the Defined Entity framework in a future releases. This would allow extensions like CSE to do object extention and reap the benefits of query engine, ACL, tenancy etc. available to vCD native objects like vApp. Under the new schema, CSE will start representing native clusters as `CSE Cluster Object` in vCD, which for all purposes will be a black box entity for vCD managed by CSE. CSE will still continue creating vApps in the background to power these `Cluster Objects`. And this new objects will host the metadata related to clusters.

On another note, when a user does a list cluster operation, that operation is translated to a vApp query in the CSE server (and the query is run in the user context). if the user has no visibility to the metadata, identifying the vApp that are cluster will become impossible. I hope that clarifies the need to keep the vApp metadata in GENERAL domain for the time being.

Regards
Aritra Sen

**Comment by ejgarcia** [comment]
Hi Aritra,

I think that using Entities in vCD is a very interesting idea, and the way to go, totally agree with that.

But regarding current usage of metadata, please take into account that I am suggesting READONLY, which is in the SYSTEM domain, but allows users to see it, query it, but not modify it. I mean, it can be queried as a regular user with

`https://<vcd>/api/query?type=vApp&fields=metadata@SYSTEM:cse.cluster.id`

and it will just return the correct value

```
    <VAppRecord href="https://<vcd>/api/vApp/vapp-7906f55d-840b-4601-a30d-cdce387b6c81" numberOfVMs="1">
        <Metadata>
            <MetadataEntry>
                <Domain visibility="READONLY">SYSTEM</Domain>
                <Key>cse.cluster.id</Key>
                <TypedValue xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:type="MetadataStringValue">
                    <Value>91300759-ebda-48d0-bd71-a39792297c62</Value>
                </TypedValue>
            </MetadataEntry>
        </Metadata>
    </VAppRecord>
```

Using READWRITE in GENERAL domain allows them to see it, query it, AND edit or remove it, which will cause problems in CSE (for example, they could change the **cse.cluster.id** to a different id, and CSE is confused... we have already tested that)

**Comment by sahithi** [comment]
Created an issue for this in Jira backlog. We will re-evaluate this. Thanks Emilio
